### ISSUE TITLE
Possible memory leak when use PooledDbContextFactory 

### ISSUE THREAD

**Issue by FarseerSky** [dotnet/efcore#32652]
Issue Description:
I have encountered a memory dump issue during my stress test. The memory dump shows that there are 95000 instances of EntityFrameworkCore.Update.ColumnModification held by 239 DbContext objects in the pool.
![image](https://github.com/dotnet/efcore/assets/11345239/ade3a8f6-a601-4d25-8b58-40e6aa77935d)

I have investigated the source code and found that in the **TopologicalSort** method of **CommandBatchPreparer**, it clears the **_modificationCommandGraph** field which stores all the commands from the last time only when a new **BatchCommands** invoked. Can this be cleared after the **SaveChangesAsync()** method instead?

Code snippet:
```C#
await using var ctx = await _contextFactory.CreateDbContextAsync();
foreach (var bean in beans)
{
    ctx.Update(bean);
}

var res = await ctx.SaveChangesAsync();
ctx.ChangeTracker.Clear();
```

System Information:

- EF Core version:8.0.0
- Database provider: Pomelo.EntityFrameworkCore.MySql 8.0.0-beta.2
- Target framework: NET 8.0
- Operating system:Linux

**Comment by MoMack20** [comment]
Will this make it into a EF Core 8 patch? If not, is the call to ctx.ChangeTracker.Clear(); enough to mitigate this issue?

**Comment by mu88** [comment]
@MoMack20 no, unfortunately, calling `context.ChangeTracker.Clear();` is not enough to mitigate the issue as you can see in [my other issue](https://github.com/dotnet/efcore/issues/33354)

**Comment by FarseerSky** [comment]
@mu88 Looking at the milestone, it should be fixed in version 9.0. I have given up using PooledDbContext, and the non-pooled performance is also very good.
